Adding the LXMERT pretraining model (MultiModal  languageXvision)  to HuggingFace's suite of models

eltoto1219 opened the pull request:
LXMERT is considered a dual-stream language-vision model. Dual stream meaning that it uses a transformer encoder to perform self-attention between each modality of data and then a cross-modality transformer encoder for a fine-grained cross-attention between each modality. It's achieved tremendous (STOTA results) success across a wide variety of downstream tasks (GQA, VQA2, NLVR2). 

Here is the original link to the paper:  https://arxiv.org/pdf/1908.07490.pdf
Here is the link to the original implementation: https://github.com/airsplay/lxmert
and here is the weight to the original model weights: https://nlp1.cs.unc.edu/data/model_LXRT.pth

Please let me know if there is anything i missed, and I would be very grateful for any help if I end up running into any blockers, but I will do my best to follow the detailed instructions in the templates. 

This is also a work in progress!

LysandreJik commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,874 @@
+            module.bias.data.zero_()
+
+
+class LxmertFeatureExtraction(LxmertPreTrainedModel):
```
We try to respect the `ModelForTask` naming scheme. In that cas it would be `LxmertForFeatureExtraction`
```suggestion
class LxmertForFeatureExtraction(LxmertPreTrainedModel):
```

LysandreJik commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,874 @@
+@add_start_docstrings(
+    """Lxmert Model with a specified pre-training heads on top. """, LXMERT_START_DOCSTRING, LXMERT_INPUTS_DOCSTRING
+)
+class LxmertPretraining(LxmertPreTrainedModel):
```
```suggestion
class LxmertForPretraining(LxmertPreTrainedModel):
```

LysandreJik commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,975 @@
+        return outputs
+
+
+class TFLxmertFeatureExtraction(TFLxmertPreTrainedModel):
```
```suggestion
class TFLxmertForFeatureExtraction(TFLxmertPreTrainedModel):
```

LysandreJik commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,975 @@
+@add_start_docstrings(
+    """Lxmert Model with a `language modeling` head on top. """, LXMERT_START_DOCSTRING, LXMERT_INPUTS_DOCSTRING
+)
+class TFLxmertPretraining(TFLxmertPreTrainedModel):
```
```suggestion
class TFLxmertForPretraining(TFLxmertPreTrainedModel):
```

eltoto1219 commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,874 @@
+logger = logging.getLogger(__name__)
+
+
+LXMERT_PRETRAINED_MODEL_ARCHIVE_LIST = {
+    "bert-base-uncased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-uncased.tar.gz",
+    "bert-large-uncased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-uncased.tar.gz",
+    "bert-base-cased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-cased.tar.gz",
+    "bert-large-cased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-cased.tar.gz",
+    "bert-base-multilingual-uncased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-multilingual-uncased.tar.gz",
+    "bert-base-multilingual-cased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-multilingual-cased.tar.gz",
+    "bert-base-chinese": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-chinese.tar.gz",
+}
```
I was meaning to update this after I added the model card. At first, I had it this way because in the paper there was a way to initialize with pretrained-bert weights on only the language modality. However, that was only done as an experiment, so I'll go ahead and just have the lxmert checkpoints here.

LysandreJik replied:
Sure, that sounds good!

eltoto1219 commented:
Thank you so much!!! I really appreciate you offering to help too!

To cover what you mentioned quickly, I can upload the model weights tomorrow! For updating the documentation, that should be no problem either. I think for the model outputs regarding this model, there probably is quite a bit of information to return (pooled output, hidden states, and attentions for the language, vision, and cross-modality encoders). I can see about adding these tomorrow and get your thoughts on that. And lastly, I am glad I can help with adding more model heads! 

I have a bit more testing to do especially with the tensorflow model, but i will see what I can get done and let you know if I run into any blockers or questions. Look forward to getting back to you!

eltoto1219 commented:
Hi, sorry for the slight delay. I added a new dataclass for Lxmert outputs, added the model card, finished all tests for the torch model, among a couple of other things. 

For the tests, I have forgone the ModelMixin parent as I have found that some of the tests are hard to apply to lxmert. I have also temporarily forgone adding example commands for lxmert. Is this a neccesity, or would it be alright to leave these out?

I think I updated the documentation to the new standard, but if there are still some errors, any help would be appreciated!

I am running a donwstream task with the pretrained weights right now to ensure that the results are still the same! I will get back to you with these in about a day!

LysandreJik commented on src/transformers/convert_pytorch_checkpoint_to_tf2.py:
```diff
@@ -181,6 +187,14 @@
     "electra": (ElectraConfig, TFElectraForPreTraining, ElectraForPreTraining, ELECTRA_PRETRAINED_CONFIG_ARCHIVE_MAP,),
+    "lxmert": (
+        LxmertConfig,
+        LxmertPretraining,
+        TFLxmertPretraining,
```
```suggestion
        LxmertForPretraining,
        TFLxmertForPretraining,
```

LysandreJik commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,975 @@
+             If set to ``True``, the visual hidden states of all layers from the cross modality encoder are returned where the visual features are used as the input and the language features are used as the context. See ``hidden_states`` under returned tensors for more detail.
+         output_l_hidden_states (:obj:`bool`, `optional`, defaults to :obj:`False`):
+             If set to ``True``, the language hidden states of all layers from the cross modality encoder are returned where the visual features are used as the context. See ``hidden_states`` under returned tensors for more detail.
+        return_dict (:obj:`bool`, `optional`, defaults to :obj:`None`):
+            If set to ``True``, the model will return a :class:`~transformers.file_utils.LxmertModelOutput` instead of a
+            plain tuple.
```
Great that you added the `return_dict`! I think it's lacking a space for indentation?

LysandreJik commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,975 @@
+        # Loss functions
+        self.loss_fcts = {
+            "l2": SmoothL1Loss(reduction="none"),
+            "visn_ce": CrossEntropyLoss(ignore_index=-1, reduction="none"),
+            "ce": CrossEntropyLoss(ignore_index=-1),
```
Previously we have `ignore_index=-1` but it brought all kinds of problems and mismatch betcause it wasn't PyTorch's default. We now try to have `ignore_index=-100` (the default) everywhere.

Would it be possible to do the same here?
```suggestion
            "visn_ce": CrossEntropyLoss(reduction="none"),
            "ce": CrossEntropyLoss(),
```

eltoto1219 replied:
I think for testing the local data, switching to -100 would lead to some problems for the local data I am using for testing, but for the future, I think once we implement a vision-language pipeline it would make much more sense, so yes I can definitely change the ignore index for this PR!

LysandreJik commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,975 @@
+        obj_labels=None,
+        matched_label=None,
+        ans=None,
+        output_v_attentions=False,
+        output_l_attentions=False,
+        output_x_attentions=False,
```
This is cool! In order to re-introduce the mixin we'll probably also have to add an argument `output_attentions` which would return all three.

LysandreJik replied:
We probably should override the attention test directly in the `test_lxmert_modeling.py` file to check these directly instead, rather than implement an additional argument which wouldn't serve much purpose.

eltoto1219 replied:
hrm, I see. I actually think for the sake of simplicity, just returning the attention from each encoder would probably be easier for the end user. However, if you do like this implementation better, I think that changing the test directly, should not be that much of a problem!

LysandreJik commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,975 @@
+        r"""
+        masked_lm_labels (``torch.LongTensor`` of shape ``(batch_size, sequence_length)``, `optional`, defaults to :obj:`None`):
+            Labels for computing the masked language modeling loss.
+            Indices should be in ``[-100, 0, ..., config.vocab_size]`` (see ``input_ids`` docstring)
```
If you want to keep the `ignore_index=-1` in the cross entropy losses, this docstring and the ones below should be updated to reflect that.

LysandreJik commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,975 @@
+}
+
+
+class TFVisualFeatEncoder(tf.keras.layers.Layer):
```
Same comment as for PyTorch

LysandreJik commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,975 @@
+"""
+
+LXMERT_INPUTS_DOCSTRING = r"""
+    Inputs:
+        **input_ids**: ``Numpy array`` or ``tf.Tensor`` of shape ``(batch_size, sequence_length)``:
+            Indices of input sequence tokens in the vocabulary.
+            To match pre-training, LXMERT input sequence should be formatted with [CLS] and [SEP] tokens as follows:
+            Indices can be obtained using :class:`transformers.LxmertTokenizer`.
+            See :func:`transformers.PreTrainedTokenizer.encode` and
+            :func:`transformers.PreTrainedTokenizer.convert_tokens_to_ids` for details.
+        **attention_mask**: (`optional`) ``Numpy array`` or ``tf.Tensor`` of shape ``(batch_size, sequence_length)``:
+            Mask to avoid performing attention on padding token indices.
+            Mask values selected in ``[0, 1]``:
+            ``1`` for tokens that are NOT MASKED, ``0`` for MASKED tokens.
+        **token_type_ids**: (`optional`) ``Numpy array`` or ``tf.Tensor`` of shape ``(batch_size, sequence_length)``:
+            Segment token indices to indicate first and second portions of the inputs.
+            Indices are selected in ``[0, 1]``: ``0`` corresponds to a `sentence A` token, ``1``
+            corresponds to a `sentence B` token
+        **visual_feats**: (`optional`) ``Numpy array`` or ``tf.Tensor`` of shape ``(batch_size, features_length, features_dim)``:
+            Pre-trained and instance-level vision features ROI-aligned and pooled from bounding boxes.
+        **visual_attention_mask**: (`optional`) ``Numpy array`` or ``tf.Tensor`` of shape ``(batch_size, features_length)``:
+            Mask to avoid performing attention on padding token indices for visual features.
+            Mask values selected in ``[0, 1]``:
+            ``1`` for tokens that are NOT MASKED, ``0`` for MASKED tokens.
```
Would be great to update to the new docstrings

LysandreJik commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,975 @@
+)
+class TFLxmertModel(TFLxmertPreTrainedModel):
+    r"""
+    Outputs: `Tuple` comprising various elements depending on the configuration (config) and inputs:
+
+        **(lang_feats, visn_feats)**: ``Tuple[Numpy array or tf.Tensor, Tuple[Numpy array or tf.Tensor]]`` of shapes
+        ``(batch_size, sequence_length, hidden_size)`` and  each element of the nested tuple being of shape
+        ``(batch_size, num_features, visual_feat_dim)`` and ``(batch_size, num_features, visual_pos_dim)`` respectively
+
+        **pooler_output**: ``Numpy array`` or ``tf.Tensor`` of shape ``(batch_size, hidden_size)``
+            Last layer hidden-state of the first token of the sequence from the cross modality encoder(classification token)
+            further processed by a Linear layer and a Tanh activation function. The Linear
```
Same here

LysandreJik commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,975 @@
+    """Lxmert Model with a `language modeling` head on top. """, LXMERT_START_DOCSTRING, LXMERT_INPUTS_DOCSTRING
+)
+class TFLxmertForPretraining(TFLxmertPreTrainedModel):
+    r"""
+    **input_ids**: ``tf.Tensor`` of shape ``(batch_size, sequence_length)``:
+        Indices of input sequence tokens in the vocabulary.
+        To match pre-training, LXMERT input sequence should be formatted with [CLS] and [SEP] tokens as follows:
+        Indices can be obtained using :class:`transformers.LxmertTokenizer`.
+        See :func:`transformers.PreTrainedTokenizer.encode` and
+        :func:`transformers.PreTrainedTokenizer.convert_tokens_to_ids` for details.
+    **attention_mask**: (`optional`) ``tf.Tensor`` of shape ``(batch_size, sequence_length)``:
+        Mask to avoid performing attention on padding token indices.
+        Mask values selected in ``[0, 1]``:
+        ``1`` for tokens that are NOT MASKED, ``0`` for MASKED tokens.
+    **token_type_ids**: (`optional`) ``tf.Tensor`` of shape ``(batch_size, sequence_length)``:
+        Segment token indices to indicate first and second portions of the inputs.
+        Indices are selected in ``[0, 1]``: ``0`` corresponds to a `sentence A` token, ``1``
+        corresponds to a `sentence B` token
+    **visual_feats**: (`optional`) ``tf.Tensor`` of shape ``(batch_size, features_length, features_dim)``:
+        Pre-trained and instance-level vision features ROI-aligned and pooled from bounding boxes.
+    **pos**: (`optional`) ``tf.Tensor`` of shape ``(batch_size, features_length, visual_pos_dim)``:
+        the bounding box coordinate for each respective visual feature instance.
+    **masked_lm_labels**: (`optional`) ``tf.Tensor`` of shape ``(batch_size, sequence_length)``:
+        Labels for computing the masked language modeling loss.
+        Indices should be in ``[-1, 0, ..., config.vocab_size]`` (see ``input_ids`` docstring)
+        Tokens with indices set to ``-100`` are ignored (masked), the loss is only computed for the tokens with labels
+        in ``[0, ..., config.vocab_size]``
+    **obj_labels**: (`optional`): ``Dict[Str: Tuple[Torch.FloatTensor, Torch.FloatTensor]]``
+        each key is named after each one of the visual losses and each element of the tuple is of the shape
+        ``(batch_size, num_features, num_objects/num_attrs)`` and ``(batch_size, num_features)``
+        for each the label id and the label score respectively
+    **matched_label**: (`optional`) ``Torch.Tensor`` of shape ``(batch_size,)`` an int [0,1] that represents
+    whether or not this text is correctly matching its associated image
+    **ans**: (`optional`) ``Torch.Tensor`` of shape ``(batch_size, num_qa_answers)`` a one hot represntation of the correct answer
```
Same here for doc

LysandreJik commented on src/transformers/tokenization_lxmert.py:
```diff
@@ -0,0 +1,438 @@
+# to pretrained vocabulary URL for all the model shortcut names.
+####################################################
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "bert-base-uncased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-uncased-vocab.txt",
+        "bert-large-uncased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-uncased-vocab.txt",
+        "bert-base-cased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-cased-vocab.txt",
+        "bert-large-cased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-large-cased-vocab.txt",
+        "bert-base-multilingual-uncased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-multilingual-uncased-vocab.txt",
+        "bert-base-multilingual-cased": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-multilingual-cased-vocab.txt",
+        "bert-base-chinese": "https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-chinese-vocab.txt",
+    }
```
Let's keep in mind to update this before merging

LysandreJik commented on src/transformers/tokenization_lxmert.py:
```diff
@@ -0,0 +1,438 @@
+    return tokens
+
+
+class LxmertTokenizer(PreTrainedTokenizer):
```
This is basically a BERT tokenizer, right? Does it have anything different? If not, you can add the fast version very simply, similarly to what is done for ELECTRA:

https://github.com/huggingface/transformers/blob/827d6d6ef071029cfe82838a18dab046b5813976/src/transformers/tokenization_electra.py#L68-L81

eltoto1219 replied:
I actually did notice it was identical to the BERTTokenizer, so I will go ahead and make this change for my next commit!

LysandreJik commented on tests/test_modeling_lxmert.py:
```diff
@@ -0,0 +1,408 @@
+    #
+
+
+class LxmertModelTester:
```
This test file is really great. It would still need to depend on the mixin to make sure we test everything.

eltoto1219 commented:
Hi thank you so much for a really quick response and review! In my next commit, Ill implement the following changes and suggestions. If it wouldn't be to much to ask,  I think I could actually do with much help adding the mixin tester. Given that there are quite a few tests that seem lxmert seems to be incompatable with (for example I think one of the tests required that the LxmertConfig had the 'num_hidden_layers' attribute, which wouldn't be applicable since I suppose instead, we let the user specify the number of hidden layers for each of the visual, language, and cross-modality encoders. You're judgement on deciding what makes sense with regards to Lxmert test compatibility is probably greater than mine. 

Also just one last implementation detail that I should probably bring up is that for the output of the cross-modality encoder attentions, I only output the attentions when the language-hidden states are used as the input to the cross attention. Since this encoder is used for bidirectional attention, I do not store the attentions when the visual-hidden states are used as the input. Also For every layer of the cross attention encoder, each of the visual and language states are further processed by a separate self-attention layer, and I do not keep track of the attention outputs for these either. The only reason I keep track of the language attentions when used as input to the cross-modality attention layer is because it is those hidden states that are used for downstream pooling. 

I have yet to change the output for the TFLxmertModeling classes, which I will probably add in the commit following the one that addresses your review, but it probably would save me some time if you were able to cover that and would be greatly appreciated aswell!

eltoto1219 commented:
Commited the above changes, I need to do some testing with the conversion script again, but besides that. I will wait to change the tensorflow code and the modeling tests for your response. If I do make a change in my next commit to modeling_tf_lxmert, it will just be to the documentation

eltoto1219 commented:
I also just added some functionality to edit the number of question answering labels, simialr to the "resize_output_embeddings" utility in the LxmertPretrainedModel class.  However, what I added seems a lot less complex than the process for resizing embeddings, so if you could take a look at these new functions, that would be awesome! I added some tests, and they do seem to work.

LysandreJik commented:
Hi! Thanks for the thorough explanation. I'm adding the common tests to the tests, and will report here. May I push directly on your fork?

LysandreJik commented:
Actually, since I'm bound to make changes to some of the files, I'd like you to review the changes I'm making so that we may discuss. I'll open a PR on your branch when it's in a good state.

eltoto1219 commented:
Okay  sounds good! the output hidden states argument should be added in
about 30 minutes, and then I can see about adding the TF changes in the
next hour.

On Mon, Aug 10, 2020 at 8:12 AM Lysandre Debut <notifications@github.com>
wrote:

> Actually, since I'm bound to make changes to some of the files, I'd like
> you to review the changes I'm making so that we may discuss. I'll open a PR
> on your branch when it's in a good state.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/transformers/pull/5793#issuecomment-671318378>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADLBORYS3OXI4RBT4LHTJLLR77P3FANCNFSM4O3KPDOA>
> .
>

LysandreJik commented:
Great, I'll take a look at doing the Mixin tomorrow once your changes are up!

eltoto1219 commented:
the lxmert model in pytorch should be ready to go!

LysandreJik commented:
Just pushed the same changes for TF as the PyTorch ones in https://github.com/eltoto1219/transformers/pull/1, alongside document changes and a few patches to the PyTorch version. 

I'll take care of the merge commit once everything is done, it's due to `isort==5` and `black==20.8b` being released.

LysandreJik commented on src/transformers/modeling_tf_pytorch_utils.py:
```diff
@@ -174,7 +174,7 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
 
         try:
-            assert list(symbolic_weight.shape) == list(array.shape)
+            assert list(symbolic_weight.shape) == list(array.shape), f"{sw_name}"
```
Remove this
```suggestion
            assert list(symbolic_weight.shape) == list(array.shape)
```

LysandreJik commented on src/transformers/configuration_lxmert.py:
```diff
@@ -0,0 +1,179 @@
+        self.visual_feat_loss = visual_feat_loss
+        self.output_hidden_states = output_hidden_states
+        self.output_attentions = self.output_attentions
+        self.num_hidden_layers = {"vision": r_layers, "cross_encoder": x_layers, "language": l_layers}
```
Opinionated choice on how to deal with `num_hidden_layers` when there are more than one encoders. This value is  used in tests for attentions/hidden states, which have been overridden to fit LXMERT.

thomwolf replied:
I'm fine with that

LysandreJik commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,1348 @@
+# coding=utf-8
+# Copyright 2018 Hao Tan, Mohit Bansal
```
Should add HuggingFace here as well

LysandreJik commented on src/transformers/modeling_auto.py:
```diff
@@ -116,6 +117,7 @@
     LongformerForTokenClassification,
     LongformerModel,
 )
+from .modeling_lxmert import LxmertForPreTraining, LxmertModel
```
The question answering model wasn't added here as it has a very different API to other QA models.

thomwolf commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,1348 @@
+            :func:`transformers.PreTrainedTokenizer.__call__` for details.
+
+            `What are input IDs? <../glossary.html#input-ids>`__
+        visual_feats: (:obj:`torch.FloatTensor` of shape :obj:՝(batch_size, num_visual_features, visual_feat_dim)՝):
+            This input represents visual features. They ROI pooled object features from bounding boxes using a faster-RCNN model)
+            These are currently not provided by the transformers library
+        visual_pos: (:obj:`torch.FloatTensor` of shape :obj:՝(batch_size, num_visual_features, visual_pos_dim)՝):
+            This input represents spacial features corresponding to their relative (via index) visual features.
+            The pre-trained lxmert model expects these spacial features to be normalized bounding boxes on a scale of 0~1.
+            These are currently not provided by the transformers library
```
Here we should add an URL or link to where people can find the missing features (or a link to the future example showing how to generate them?)

thomwolf reviewed (approved):
This is really great!
I'm fine with the hidden-states and attention layers.

sgugger commented on docs/source/model_doc/lxmert.rst:
```diff
@@ -0,0 +1,110 @@
+Overview
+~~~~~~~~~~~~~~~~~~~~~
+
+The LXMERT model was proposed in `LXMERT: Learning Cross-Modality Encoder Representations from Transformers`
```
We should add a link to the paper here.

sgugger commented on docs/source/model_doc/lxmert.rst:
```diff
@@ -0,0 +1,110 @@
+
+Tips:
+
+- Bounding boxes are not necessary to be used in the visual feature embeddings, any kind of visual-spacial features will work
```
```suggestion
- Bounding boxes are not necessary to be used in the visual feature embeddings, any kind of visual-spacial features will work.
```

sgugger commented on docs/source/model_doc/lxmert.rst:
```diff
@@ -0,0 +1,110 @@
+- The bi-directional cross-modality encoder attention only returns attention values when the language modality is used as the input and the vision modality is used as the context vector. Further,
+  while the cross-modality encoder contains self-attention for each respective modality and cross-attention, only the cross attention is returned and both self attention outputs are disregarded.
+
+The code can be found `https://github.com/airsplay/lxmert`
```
```suggestion
The code can be found `here <https://github.com/airsplay/lxmert>`__.
```

sgugger commented on docs/source/pretrained_models.rst:
```diff
@@ -363,4 +363,8 @@ For a list that includes community-uploaded models, refer to `https://huggingfac
\ No newline at end of file
++-------------------+------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------+
+| Lxmert            | ``lxmert-base-uncased``                                    | | 9-language layers, 9-relationship layers, and 12-cross-modality layers                                                              |
+|                   |                                                            | | 768-hidden, 12-heads (for each layer) ~XXXM parameters                                                                              |
```
We should remove the XXX comment or put an actual number.

sgugger commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,1348 @@
+    """
+
+    question_answering_score: torch.FloatTensor
+    loss: Optional[torch.FloatTensor] = None
```
loss should be the first here, to be consistent with other model outputs. It's ok to have `question_answering_score` default to None even if it's a require parameter.

sgugger commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,1348 @@
+
+
+LXMERT_START_DOCSTRING = r"""
+    The LXMERT model was proposed in `LXMERT: Learning Cross-Modality Encoder Representations from Transformers`
```
Add a link to the article here.

sgugger commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,1348 @@
+        pooled_output = self.pooler(lang_output)
+
+        if not return_dict:
+            return (lang_output, visual_output, pooled_output) + hidden_states + all_attentions
```
Very clean!

sgugger commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,1348 @@
+        return outputs
+
+
+class LxmertSelfattLayer(nn.Module):
```
nit: I think `LxmertSelfAttentionLayer` would be more consistent with our usual naming patterns.

sgugger commented on src/transformers/modeling_lxmert.py:
```diff
@@ -0,0 +1,1348 @@
+        )
+
+
+class LxmertVisualFeatEncoder(nn.Module):
```
nit: I think `LxmertVisualFeatureEncoder` would be more consistent with our usual naming patterns.

sgugger commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,1248 @@
+    cross_encoder_attentions: Optional[Tuple[tf.Tensor]] = None
+
+
+class TFLxmertVisualFeatEncoder(tf.keras.layers.Layer):
```
nit: I think `TFLxmertVisualFeatureEncoder` would be more consistent with our usual naming patterns.

sgugger commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,1248 @@
+        return hidden_states
+
+
+class TFLxmertAttOutput(tf.keras.layers.Layer):
```
nit: I think `TFLxmertAttentionOutput` would be more consistent with our usual naming patterns.

sgugger commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,1248 @@
+        return hidden_states
+
+
+class TFLxmertSelfattLayer(tf.keras.layers.Layer):
```
nit: I think `TFLxmertSelfAttentionLayer` would be more consistent with our usual naming patterns.

sgugger commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,1248 @@
+        return (attention_output, attention_probs) if output_attentions else (attention_output,)
+
+
+class TFLxmertCrossattLayer(tf.keras.layers.Layer):
```
nit: I think `TFLxmertCrossAttentionLayer` would be more consistent with our usual naming patterns.

sgugger commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,1248 @@
+
+
+LXMERT_START_DOCSTRING = r"""
+    The LXMERT model was proposed in `LXMERT: Learning Cross-Modality Encoder Representations from Transformers`
```
Add a link to the article here.

sgugger commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,1248 @@
+    using a combination of masked language modeling, region of interest feature regression,
+    cross entropy loss for question answering attribute prediction, and object tag predicition.
+
+    This model is a tf.keras.Model `tf.keras.Model` sub-class. Use it as a regular TF 2.0 Keras Model and
```
Add a link to tf.keras.Model doc, as in the other TF models.

sgugger commented on src/transformers/modeling_tf_lxmert.py:
```diff
@@ -0,0 +1,1248 @@
+
+LXMERT_INPUTS_DOCSTRING = r"""
+    Args:
+        input_ids (:obj:`Numpy array` or :obj:`tf.Tensor` of shape :obj:`(batch_size, sequence_length)`):
```
```suggestion
        input_ids (:obj:`np.ndarray` or :obj:`tf.Tensor` of shape :obj:`(batch_size, sequence_length)`):
```
